### PR TITLE
Ensure capacity freed by clear is made available to GKOP borrowers. J…

### PR DESCRIPTION
**Pull request by psteitz**
…IRA: POOL-391

### CODE REVIEW

File: src/main/java/org/apache/commons/pool2/impl/GenericKeyedObjectPool.java [hunk 1]
```diff
@@ -570,23 +569,47 @@ private int calculateDeficit(final ObjectDeque<T> objectDeque) {
...
     @Override
     public void clear() {
-        poolMap.keySet().forEach(this::clear);
-    }
+        final Iterator<K> iter = poolMap.keySet().iterator();
 
+        while (iter.hasNext()) {
+            clear(iter.next(), false);
+        }
+    }
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/pool2/impl/GenericKeyedObjectPool.java:577]: Why not use a lambda?

Comment (psteitz) [reply src/main/java/org/apache/commons/pool2/impl/GenericKeyedObjectPool.java:573]: Done

File: src/test/java/org/apache/commons/pool2/impl/TestGenericKeyedObjectPool.java [hunk 2]
```diff
@@ -2474,6 +2474,138 @@ public void testWhenExhaustedBlockClosePool() throws Exception {
         assertTrue(wtt.thrown instanceof InterruptedException);
     }
 
+    /**
+     * POOL-391
+     * Adapted from code in the JIRA ticket.
+     *
+     * @throws Exception May occur in some failure modes
+     */
+    @Test
```
Comment (kinow) [inline src/test/java/org/apache/commons/pool2/impl/TestGenericKeyedObjectPool.java:2483]: Great job on using the code from the JIRA for the test :tada: 

Unfortunately I think there is something wrong with the formatting here :point_down:

File: src/test/java/org/apache/commons/pool2/impl/TestGenericKeyedObjectPool.java [hunk 3]
```diff
@@ -2474,6 +2474,138 @@ public void testWhenExhaustedBlockClosePool() throws Exception {
...
+    public void testInvalidateFreesCapacityForOtherKeys() throws Exception {
+        gkoPool.setMaxTotal(1);
+        gkoPool.setMaxWait(Duration.ofMillis(500));
+        Thread borrower = new Thread(new SimpleTestThread<>(gkoPool, "two"));
+        String obj = gkoPool.borrowObject("one");
+        borrower.start();  // Will block
+        Thread.sleep(100);  // Make sure borrower has started
+        gkoPool.invalidateObject("one", obj);  // Should free capacity to serve the other key
+        Thread.sleep(20);  // Should have been served by now
+        assertFalse(borrower.isAlive());
```
Comment (kinow) [inline src/test/java/org/apache/commons/pool2/impl/TestGenericKeyedObjectPool.java:2606]: I started trying to rely less on `Thread.sleep` and use more mocks or libraries that handle the timeouts (been doing more Python & JS, but found this one in Java that's similar to what I'm using https://github.com/awaitility/awaitility). But probably for a follow-up issue/discussion :+1:

**Review by kinow — APPROVED** [review]
Code looks OK to me, easy to follow and with good tests! Thanks @psteitz ! Added one minor comment about the code formatting in the test. Bruno.

**Pull request merged**